epoch.go

```diff
@@ -1534,22 +1574,46 @@ func (e *Epoch) handleBlockMessage(message *BlockMessage, from NodeID) error {
 		msgForRound.proposalBeingProcessed = true
 	}
 
-	// Create a task that will verify the block in the future, after its predecessors have also been verified.
-	task := e.createBlockVerificationTask(e.oneTimeVerifier.Wrap(block), from, vote)
+	return nil
+}
 
-	// isBlockReadyToBeScheduled checks if the block is known to us either from some previous round,
-	// or from storage. If so, then we have verified it in the past, since only verified blocks are saved in memory.
-	canBeImmediatelyVerified := e.isBlockReadyToBeScheduled(md.Seq, md.Prev)
+// blockDependencies returns the dependencies bh has before it can be verified.
+// It returns the digest of the previous block it depends on (or emptyDigest if none),
+// as well as a list of rounds for which it needs to verify empty notarizations.
+// TODO: we should request empty notarizations if we don't have them
+func (e *Epoch) blockDependencies(bh BlockHeader) (*Digest, []uint64) {
+	if bh.Seq == 0 {
+		// genesis block has no dependencies
+		return nil, nil
+	}
 
-	// Schedule the block to be verified once its direct predecessor have been verified,
-	// or if it can be verified immediately.
-	e.Logger.Debug("Scheduling block verification",
-		zap.Uint64("round", md.Round),
-		zap.Uint64("seq", md.Seq),
-		zap.Bool("ready", canBeImmediatelyVerified))
-	e.sched.Schedule(task, md.Prev, canBeImmediatelyVerified)
+	prevBlockDependency := &bh.Prev
 
-	return nil
+	prevBlock, notarizationOrFinalization, found := e.locateBlock(bh.Seq-1, bh.Prev[:])
+	if !found {
+		// should never happen since we check this when we verify the proposal metadata
+		e.Logger.Error("Could not find predecessor block for proposal scheduling",
```

Comment:
Can you explain why we won't run into this problem when we replicate notarized blocks?